src/Features/Core/Completion/MostRecentlyUsedList.cs

```diff
@@ -0,0 +1,58 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.CodeAnalysis.Completion
+{
+    internal class MostRecentlyUsedList<T>
+    {
+        private readonly object gate = new object();
+        private readonly List<T> items = new List<T>();
```

Comment:
You're right. You should be implementing an ICompletionRules, since that's what they're designed for.